Allow samplers to modify tracestate (#988)
open-telemetry/opentelemetry-specification: specification/trace/sdk.md

  * `RECORD_AND_SAMPLE` - `IsRecording() == true` AND `Sampled` flag` MUST be set.
* A set of span Attributes that will also be added to the `Span`. The returned
object must be immutable (multiple calls may return different immutable objects).
* A `Tracestate` that will be associated with the `Span` through the new
  `SpanContext`. If non-empty `Tracestate` was passed in `SpanContext`,

>   it MUST be populated on the `SamplingResult` even if tracestate was not

Reviewer: > If non-empty `Tracestate` was passed in `SpanContext`

Is the SpanContext passed as argument to shouldSample meant here?

> it MUST be populated

What is meant by "it must be populated"?

Can we just say. "Note: If the sampler returns an empty tracestate here, the tracestate will be cleared, so samplers should normally return the passed-in tracestate here if they do not intent to change it."

  * `RECORD_AND_SAMPLE` - `IsRecording() == true` AND `Sampled` flag` MUST be set.
* A set of span Attributes that will also be added to the `Span`. The returned
object must be immutable (multiple calls may return different immutable objects).
* A `Tracestate` that will be associated with the `Span` through the new
  `SpanContext`.
  Note: If the sampler returns an empty `Tracestate` here, the `Tracestate` will be cleared,

>   so samplers should normally return the passed-in `Tracestate` if they do not intend

Reviewer 2: ```suggestion
  so samplers SHOULD return the `Tracestate` passed to them, if they do not intend
```

Reviewer: In a Note, we should not have upper-case requirements. If you want reword it, also drop the "Note:"

  * `RECORD_AND_SAMPLE` - `IsRecording() == true` AND `Sampled` flag` MUST be set.
* A set of span Attributes that will also be added to the `Span`. The returned
object must be immutable (multiple calls may return different immutable objects).

> * A `Tracestate` that will be assigned to `Span`. If non-empty `Tracestate` was

Reviewer 3: I don't think it makes sense to say that the tracestate will be "assigned" to the span. The TraceState lives in the SpanContext, not the Span itself, correct?

Reviewer 4: "Associated"?

Author: @iNikem

> By [W3C Trace Context specification](https://www.w3.org/TR/trace-context/#tracestate-header), `Tracestate` is immutable. So is `SpanContext`. How can samplers change immutable object inside another immutable object?

Tracestate can be [mutated](https://www.w3.org/TR/trace-context/#mutating-the-tracestate-field). 

This change enables folloing scenario
- propagator parses trace-context (with tracestate)
- sampler is invoked
   - it can read a parent tracestate (possible today)
   - make a copy and modify it 
   - add it to sampling result (not possible today)
- open-telemetry sdk creates a new `SpanContext` with a tracestate from sampling result (not from parent context)

Let me know if this helps.
